src/evaluator/function/logical.cpp

```diff
@@ -0,0 +1,143 @@
+#include "../../ast/kind.h"
+#include "../error.h"
+#include "../function.h"
+
+namespace evaluator {
+
+using ast::Element;
+using ast::ElementKind;
+
+std::shared_ptr<Element> AndFunction::call(CallFrame frame) const {
+    if (frame.arguments.size() != 2) {
+        throw EvaluationError(
+            "`and` expects 2 arguments, received " +
+                std::to_string(frame.arguments.size()),
+            frame.call_site
+        );
+    }
+
+    auto a_element = frame.arguments[0];
+    auto b_element = frame.arguments[1];
+
+    if (a_element->kind != ElementKind::BOOLEAN ||
+        b_element->kind != ElementKind::BOOLEAN) {
+        throw EvaluationError(
+            "`and` expects arguments to be booleans", frame.call_site
+        );
+    }
+
+    bool result;
+
+    auto a_bool = std::dynamic_pointer_cast<ast::Boolean>(a_element);
+    auto b_bool = std::dynamic_pointer_cast<ast::Boolean>(b_element);
+    result = a_bool->value && b_bool->value;
```

Comment:
```suggestion
    auto a_bool = std::dynamic_pointer_cast<ast::Boolean>(a_element);
    auto b_bool = std::dynamic_pointer_cast<ast::Boolean>(b_element);
    auto result = a_bool->value && b_bool->value;
```

ну и аналогично в остальных функциях